src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -305,71 +280,204 @@ public String decrypt(String value) throws GeneralSecurityException, IOException
+     * 
+     * @return
+     * @throws GeneralSecurityException
+     * @throws IOException
+     */
+    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+    private synchronized KeyPair getKeyPairFromAndroidKeyStoreForEncryption() throws GeneralSecurityException, IOException {
+        final KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
+        keyStore.load(null);
+        // For encryption purpose, if alias does not exist in AndroidKeyStore, 
+        // generate a new keypair. 
+        if (!keyStore.containsAlias(KEY_STORE_CERT_ALIAS)) {
+            Logger.v(TAG, "Key entry is not available, generating new entry.");
+            Calendar start = Calendar.getInstance();
+            Calendar end = Calendar.getInstance();
+            end.add(Calendar.YEAR, 100);
 
-        // Set to encrypt mode
-        Cipher cipher = Cipher.getInstance(CIPHER_ALGORITHM);
-        Mac mac = Mac.getInstance(MAC_ALGORITHM);
-        cipher.init(Cipher.ENCRYPT_MODE, mKey, ivSpec);
+            // self signed cert stored in AndroidKeyStore to asym. encrypt key
+            // to a file
+            final KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA",
+                    "AndroidKeyStore");
+            generator.initialize(getKeyPairGeneratorSpec(mContext, start.getTime(), end.getTime()));
+            generator.generateKeyPair();
+            Logger.v(TAG, "Key entry is generated");
+        } else {
+            Logger.v(TAG, "Key entry is available");
+        }
 
-        byte[] encrypted = cipher.doFinal(bytes);
+        // Read key pair again
+        return readKeyPairFromKeyEntry(keyStore);
+    }
+    
+    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+    private final synchronized SecretKey readSecretKeyFromKeyStoreForDecryption() throws GeneralSecurityException, IOException {
+        if (mSecretKeyFromAndroidKeyStore != null) {
+            return mSecretKeyFromAndroidKeyStore;
+        }
 
-        // Mac output to sign encryptedData+IV. Keyversion is not included
-        // in the digest. It defines what to use for Mac Key.
-        mac.init(mMacKey);
-        mac.update(blobVersion);
-        mac.update(encrypted);
-        mac.update(iv);
-        byte[] macDigest = mac.doFinal();
+        final File keyFile = new File(mContext.getDir(mContext.getPackageName(), Context.MODE_PRIVATE),
+                ADALKS);
+        if (mKeyPair == null) {
+            final KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
+            keyStore.load(null);
+            if (!keyStore.containsAlias(KEY_STORE_CERT_ALIAS)) {
+                Logger.v(TAG, "Key store alias is not avaliable.");
+                // Due to an issue with android keystore, when user change lockscreen type, 
+                // key data and alias could be wiped. Key data will be unrecoverable if being 
+                // wiped out.
+                throw new UnrecoverableKeyException("AndroidKeyStore does not contain the alias.");
```

Comment:
How is this exception surfaced to the app developer? What are they supposed to do when they see this exception?